### PR TITLE
Updates regarding doctrine performances with associations

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ∅
| License       | MIT
| Doc PR        | ∅

- [x] Do not join twice when filtering on an association (`EagerLoadingExtension` + `SearchFilter`) (tested)
- [x] Inner join instead of left join relations when those aren't nullable `EagerLoadingExtension` (tested)
- [x] Do not to fetch properties that are not serializable! (tested)
- [x] Discuss `Query::HINT_FORCE_PARTIAL_LOAD` to avoid fetching OneToOne relations when we don't need them

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 1]
```diff
@@ -354,4 +367,21 @@ private function normalizeValues(array $values) : array
 
        return array_values($values);
    }
+
+    private function hasJoin($queryBuilder, $alias, $association)
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: Missing type hints.

Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: By the way I think `getJoin` returning `null` when the join doesn't exist would be a cleaner API.

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 2]
```diff
@@ -101,8 +101,15 @@ public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $q
                 $parentAlias = $alias;
 
                 foreach ($propertyParts['associations'] as $association) {
-                    $alias = $queryNameGenerator->generateJoinAlias($association);
-                    $queryBuilder->join(sprintf('%s.%s', $parentAlias, $association), $alias);
+                    $join = $this->hasJoin($queryBuilder, $alias, $association);
+
+                    if (false === $join) {
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: Can't all the following block moved in the private function? It looks duplicated.

**Comment by dunglas** [comment]
Great improvements!

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php (hunk 2 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: +1 though the same

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 3]
```diff
@@ -39,23 +51,55 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
+            $method = false === $mapping['joinColumns'][0]['nullable'] ? 'innerJoin' : 'leftJoin';
+
+            $associationAlias = $relationAlias.$i;
+            $queryBuilder->{$method}($originAlias.'.'.$association, $associationAlias);
+            $select = [];
+            $targetClassMetadata = $queryBuilder->getEntityManager()->getClassMetadata($mapping['targetEntity']);
+
+            //@TODO useless at the moment, but Doctrine queries non-necessary relations because addSelect($associationAlias) lazy fetch every association
+            //By using this array of properties, it Lazy fetch association, which are initially in an INNER JOIN (fetches them twice ?!)
+            //If I try to use associationAlias when it has every property, but the properties array when it doesn't it fails because it can not fetch parent association
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: This is still an issue, @Simperfit mind taking a look? I think you've a good background with Doctrine maybe you can help figuring this out :/.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 4]
```diff
@@ -39,23 +51,55 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
+    }
+
     /**
      * Left joins relations to eager load.
      *
      * @param QueryBuilder $queryBuilder
      * @param string       $resourceClass
      */
-    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass)
+    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, string $originAlias = 'o', string &$relationAlias = 'a')
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Warning it's kind of messy actually, still trying to figure out how to do this properly.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 5]
```diff
@@ -39,23 +51,55 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
-                $queryBuilder->leftJoin('o.'.$association, 'a'.$i);
-                $queryBuilder->addSelect('a'.$i);
+            if (ClassMetadataInfo::FETCH_EAGER !== $mapping['fetch'] || false === $propertyMetadata->isReadableLink()) {
+                continue;
             }
+
+            $method = false === $mapping['joinColumns'][0]['nullable'] ? 'innerJoin' : 'leftJoin';
+
+            $associationAlias = $relationAlias.$i;
+            $queryBuilder->{$method}($originAlias.'.'.$association, $associationAlias);
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: This is a great improvement and can save lots of unnecessary left joins :).

File: src/Bridge/Doctrine/Orm/Extension/OrderExtension.php [hunk 6]
```diff
@@ -37,6 +37,7 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
         $classMetaData = $queryBuilder->getEntityManager()->getClassMetadata($resourceClass);
         $identifiers = $classMetaData->getIdentifier();
 
+        //@TODO Why?
         if (null !== $this->order) {
             foreach ($identifiers as $identifier) {
                 $queryBuilder->addOrderBy('o.'.$identifier, $this->order);
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/OrderExtension.php]: @dunglas Filter order will get on top of this but I was just wondering why is this necessary? Thanks

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 7]
```diff
@@ -88,9 +87,13 @@ private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass
...
+            //@TODO useless at the moment, but Doctrine queries non-necessary relations because addSelect($associationAlias) will try to fetch every
+            //association belonging to the association, but if those are not serialized we don't need them ! Can be especially anoying when
+            //the relation is a reverse relation (so doctrine does lots of subqueries). Here I need a way to tell doctrine to take only what we need.
+            //
+            //By using this array of properties, Doctrine will lazy fetch every associations, which are initially in an INNER JOIN (fetches them twice ?!)
+            //If I try to use associationAlias when it has every property (so that Doctrine does not do sub queries on top of the initial one) it fails because it can not fetch parent association (like internally they try to find aliasX, but no aliasX is selected, we've instead aliasX.id, aliasX.name)
+            //I obviously tried to specify `WITH` `aliasX.id = aliasY.id` without success
             foreach ($this->getMetadataProperties($mapping['targetEntity']) as $property => $propertyMetadata) {
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Please read my story, if you have any clue or idea to help me think it'd be wonderful :D. ping @Simperfit

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Note to myself:
use `partial` [ref](http://stackoverflow.com/questions/9465353/doctrine-wont-let-me-select-specific-fields)

File: src/Bridge/Doctrine/Orm/Extension/OrderExtension.php (hunk 6 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/OrderExtension.php]: This is the default ORDER BY.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: We need a central point of reference for the joins already made, so that there will be no repeated joins made by `EagerLoadingExtension` and `SearchFilter`. And I don't see why we should have default values for these parameters?

File: src/Bridge/Doctrine/Orm/Extension/OrderExtension.php (hunk 6 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/OrderExtension.php]: This I understood, but why do we need one?

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: This is done through the DQL, it's the most accurate solution IMO because it also checks for joins added through user-made filters for example.
Default values are to avoid repeating those in both applyToItem and applyToCollection methods but I can rectify this.

File: src/Bridge/Doctrine/Orm/Extension/OrderExtension.php (hunk 6 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/OrderExtension.php]: We want results ordering to be deterministic. If no `ORDER BY` is specified in the SQL, the database is free to return rows in any order.

File: tests/Bridge/Doctrine/Orm/Extension/EagerLoadingExtensionTest.php [hunk 8]
```diff
@@ -131,16 +131,16 @@ public function testApplyToItem()
         $classMetadataProphecy->getAssociationNames()->shouldBeCalled()->willReturn([0 => 'relatedDummy', 'relatedDummy2']);
         $classMetadataProphecy->associationMappings = [
             'relatedDummy' => [
-                'fetch' => 3, 'joinColumns' => [['nullable' => true]], 'targetEntity' => DummyRelated::class
+                'fetch' => 3, 'joinColumns' => [['nullable' => true]], 'targetEntity' => DummyRelated::class,
```
Comment (teohhanhui) [inline tests/Bridge/Doctrine/Orm/Extension/EagerLoadingExtensionTest.php]: 1 on each line please...

File: src/Bridge/Doctrine/Orm/CollectionDataProvider.php [hunk 9]
```diff
@@ -67,6 +68,9 @@ public function getCollection(string $resourceClass, string $operationName = nul
             }
         }
 
-        return $queryBuilder->getQuery()->getResult();
+        $query = $queryBuilder->getQuery();
+        //forces doctrine to not lazy load relations, see http://stackoverflow.com/questions/9848747/primary-key-of-owning-side-as-a-join-column
+        $query->setHint(Query::HINT_FORCE_PARTIAL_LOAD, true);
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: I think that this might give unwanted results, so I was wondering what solutions we might have. 
Maybe that just documenting it would be enough, but as this is only modifiable on the `Query` object (not the `QueryBuilder`), we may want to add a easy way to set the option. WDYT?

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: > This is done through the DQL, it's the most accurate solution IMO because it also checks for joins added through user-made filters for example.

If that's the case, then what do you mean by:

> Do not join twice when filtering on an association

**Comment by soyuka** [comment]
Ready for review :D ! 

This improves the complex queries a lot! Strong impact on performances, mainly on collections with lots of related fields (almost counts in seconds).

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: How does Doctrine detect duplicate joins?

File: src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php [hunk 10]
```diff
@@ -207,4 +206,53 @@ protected function addJoinsForNestedProperty(string $property, string $rootAlias
...
+     * Adds a join to the queryBuilder if none exists.
+     *
+     * @param QueryBuilder       $queryBuilder
+     * @param QueryNameGenerator $queryNameGenerator
+     * @param string             $alias
+     * @param string             $association        the association field
+     *
+     * @return string the new association alias
+     */
+    protected function addJoinOnce($queryBuilder, $queryNameGenerator, $alias, $association): string
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: Why `protected`?

Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: Missing typehints btw

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 4 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: It doesn't.

EagerLoadingExtension is effective before FilterExtension. If a join has been done there already, it was doing the join twice. Joining once is done in [`AbstractFilter`](https://github.com/api-platform/core/pull/717/commits/b9e04b451328bd2a4c1e4285c518dc43432d2975#diff-756b3787605765d5c843f78c389e03feR244).

File: src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php (hunk 10 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: Protected because it's called from filters (ie `SearchFilter`).

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 11]
```diff
@@ -98,12 +98,8 @@ public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $q
             if ($this->isPropertyNested($property)) {
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: @dunglas in fact this might call the `addJoinsForNestedProperties` method from the `AbstractFilter`, should I change this as well? We can then set `private` on `addJoinOnce`.

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: I've done this, removes a few lines in the `SearchFilter` but it still uses `addJoinOnce` on its own (l 153).

**Comment by dunglas** [comment]
Is it ready to be merged?

File: src/Bridge/Doctrine/Orm/CollectionDataProvider.php (hunk 9 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: @dunglas we may need to review this before merging I'm not sure that everyone would want this behavior. Wdyt?

**Comment by soyuka** [comment]
Apart from my comment on query hint it's good to go.

File: src/Bridge/Doctrine/Orm/CollectionDataProvider.php (hunk 9 above)
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: Maybe can we allow to change this trough a parameter the constructor?

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: Yes we should, but I'm not sure that a configuration parameter is the best
for this as it might be interesting to allow the configuration on the whole
"Query" object.
Anyway I'll add a parameter for now to skip lazy loading.

Comment (dunglas) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: The other way is to add a `public getResult(Query $query)` method. It will allow users to decorate the class and change the behavior of this method.

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/CollectionDataProvider.php]: I knew you had a solution, thanks I'm implementing this asap.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 12]
```diff
@@ -39,23 +50,59 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
+
+                //the field test allows to add methods to a Resource which do not reflect real database fields
+                if (true === $targetClassMetadata->hasField($property) && true === $propertyMetadata->isReadable()) {
+                    $select[] = $property;
+                }
             }
+
+            $queryBuilder->addSelect(sprintf('partial %s.{%s}', $associationAlias, implode(',', $select)));
+
+            $relationAlias = $relationAlias.++$j;
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: is there any reason that $j could not be $nbOfJoin ?

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php [hunk 13]
```diff
@@ -96,19 +96,8 @@ public function apply(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $q
...
-                foreach ($propertyParts['associations'] as $association) {
-                    $alias = $queryNameGenerator->generateJoinAlias($association);
-                    $queryBuilder->join(sprintf('%s.%s', $parentAlias, $association), $alias);
-                    $parentAlias = $alias;
-                }
-
-                $field = $propertyParts['field'];
-
-                $metadata = $this->getNestedMetadata($resourceClass, $propertyParts['associations']);
+                list($alias, $field, $associations) = $this->addJoinsForNestedProperty($property, $alias, $queryBuilder, $queryNameGenerator);
```
Comment (Simperfit) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: @dunglas There is another thing that may be better than list but I don't remember what it is, I've seen it in a PR in symfony

**Comment by Simperfit** [comment]
@soyuka Amazing work ;)

File: src/Bridge/Doctrine/Orm/Filter/SearchFilter.php (hunk 13 above)
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/SearchFilter.php]: The splat operator?

My preference go to returning an array with explicit keys and use the keys in the calling method.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 12 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Yeah it's not really the number of join, it's a way to keep the alias unique across relation, ie a random number.

**Comment by soyuka** [comment]
@dunglas I added a decoration possibility, not sure if I have to catch an undefined method call though. Should I add the same for the ItemDataProvider? Should I add an interface to be implemented by decorators using `getQuery`?

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 12 above)
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: We should avoid random numbers in DQL as much as possible to trigger the Doctrine cache as often as possible.

**Comment by dunglas** [comment]
@soyuka I would have done it the other way: just declare `public function getQuery(QueryBuilder $queryBuilder)` in this class but don't add `$decorated` support. The goal here is to let the user decorating this class, not to decorate another one.
And no need for an interface here IMO.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 12 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: It's the same in every query ofc.

**Comment by soyuka** [comment]
Done thanks. I've added the same on `ItemDataProvider` for the sake of consistency.

**Comment by dunglas** [comment]
@soyuka After double checking, isn't this use case already supported by the `QueryResultExtensionInterface` introduced by @sroze?

Someone wanting to set `Query::HINT_FORCE_PARTIAL_LOAD` can register an extension having a `getResult` method doing it. WDYT?

**Comment by soyuka** [comment]
What I don't like with this interface is that it expects the query to return a result. For example, if you wanted to use the `PaginationExtension` but also modify some hints on `Query` it won't be possible. Thing is, `Query` can only be modified only once every `QueryBuilder` tasks are done...

**Comment by dunglas** [comment]
But for such cases, decoration can be used (decorating the `PaginationExtension`, and especially its `getResult` method for instance).

**Comment by soyuka** [comment]
Indeed, this is what I was going to say. For example:

```
    /**
     * {@inheritdoc}
     */
    public function getResult(QueryBuilder $queryBuilder)
    {
        $query = $queryBuilder->getQuery();
        $query->setHint(Query::HINT_FORCE_PARTIAL_LOAD, true);

        $doctrineOrmPaginator = new DoctrineOrmPaginator($query, $this->useFetchJoinCollection($queryBuilder));
        $doctrineOrmPaginator->setUseOutputWalkers($this->useOutputWalkers($queryBuilder));

        return new Paginator($doctrineOrmPaginator);
    }
```

**Comment by soyuka** [comment]
Wait for tests to pass and to me it's good to go.

**Comment by dunglas** [comment]
So, can you revert the last commit introducing the `$decorated` constructor parameter?

File: src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php [hunk 14]
```diff
@@ -188,23 +188,72 @@ protected function extractProperties(Request $request) : array
...
+    }
+
+    /**
+     * Get the existing join from queryBuilder DQL parts.
+     *
+     * @param QueryBuilder $queryBuilder
+     * @param string       $alias
+     * @param string       $association  the association field
+     *
+     * @return Doctrine\ORM\Query\Expr\Join|null
```
Comment (dunglas) [inline src/Bridge/Doctrine/Orm/Filter/AbstractFilter.php]: Can you add a `use` statement instead?

**Comment by soyuka** [comment]
Did this already :).

**Comment by dunglas** [comment]
Thanks @soyuka
